_posts/2023-07-24-native-startup-rss-troubleshooting.adoc

```diff
@@ -0,0 +1,228 @@
+...
+2023-07-21 10:13:11,304 INFO  [io.quarkus] (main) getting-started 1.0.0-SNAPSHOT native (powered by Quarkus 2.7.6.Final) started in 0.023s. Listening on: http://0.0.0.0:8080
+
+$ ps -p $(pidof getting-started-1.0.0-SNAPSHOT-runner) -o rss=
+33024
+----
+
+One of the Quarkus versions that presented the increase in RSS consumption was 2.13.4.Final.
+In our environment, the RSS consumption had gone up to about ~36M, roughly a 10% increase:
+
+[source,bash]
+----
+$ getting-started/target/getting-started-1.0.0-SNAPSHOT-runner -Xmx64m
+...
+2023-07-21 10:24:38,529 INFO  [io.quarkus] (main) getting-started 1.0.0-SNAPSHOT native (powered by Quarkus 2.13.4.Final) started in 0.027s. Listening on: http://0.0.0.0:8080
+
+$ ps -p $(pidof getting-started-1.0.0-SNAPSHOT-runner) -o rss=
+36608
+----
+
+Early on we did some investigation to see if the GraalVM version used had an impact but it did not.
+In other words, using the same GraalVM version to build both versions resulted in similar differences.
+
+Profiling the startup using the JVM mode is something we also tried,
+and we could observe some slight increase in heap usage,
+but in JVM mode, the RSS increase was not apparent at runtime.
+So, the challenge for us was to make sense of the RSS increase that seemed to only affect native mode,
+and that was specific to the Quarkus code.
+
+Following
+https://www.brendangregg.com/FlameGraphs/memoryflamegraphs.html[Brendan Gregg’s Memory Leak (and Growth) Flame Graphs guide] on a Linux environment,
+we were able to narrow down the issue to the shape of the flamegraphs for the `mmap` system calls.
+To obtain the flamegraphs, we capture the `sys_enter_mmap` system call events with `perf record`,
+and then we generate flamegraphs with the recorded data:
+
+[source,bash]
+----
+$ sudo perf record -e syscalls:sys_enter_mmap --call-graph dwarf -- getting-started/target/getting-started-1.0.0-SNAPSHOT-runner -Xmx64m
+...
+2023-07-21 10:13:11,304 INFO  [io.quarkus] (main) getting-started 1.0.0-SNAPSHOT native (powered by Quarkus 2.7.6.Final) started in 0.023s. Listening on: http://0.0.0.0:8080
+
+$ perf script -i perf.data > out.stacks
+$ /opt/FlameGraph/stackcollapse-perf.pl < out.stacks | /opt/FlameGraph/flamegraph.pl \
+   --color=mem \
+   --title="mmap Flame Graph" \
+   --countname="calls" > out.svg
+----
+
+[NOTE]
+====
+The native executable above was created passing in
+`-Dquarkus.native.debug.enabled` and `-Dquarkus.native.additional-build-args=-H:-DeleteLocalSymbols`
+as extra build flags.
+====
+
+The areas of interest in the flamegraphs are those stacktraces that contain `ApplicationImpl.doStart()` method call.
+If you focus the flamegraph on that method,
+and then click on `Search` at the top right and type `allocate`,
+you can observe which of the `sys_enter_mmap` system calls are related to native heap memory allocation.
+These allocations are for heap chunks, each of which is 1MB,
```

Comment:
Maybe its worth mentioning it's no longer 1MB and has since been changed to 512KB

Reply:
Added a small comment around the upcoming change to reduce it.